src/test/sql/sqlite_testrunner/sqlite_testrunner.cpp

```diff
@@ -51,21 +50,78 @@ class SQLiteTestRunner : public BaseTestWithParam<std::string> {
       std::string table_file = args.at(0);
       std::string table_name = args.at(1);
 
-      DebugAssert(!StorageManager::get().has_table(table_name), "Table already loaded");
+      // Store loaded tables in a map that basically caches the loaded tables. In case the table
+      // needs to be reloaded (e.g., due to modifications), we also store the file path.
+      _tables_to_test_map.emplace(table_name, TestTable{load_table(table_file, 10), table_file});
 
-      _sqlite->create_table_from_tbl(table_file, table_name);
-
-      std::shared_ptr<Table> table = load_table(table_file, 10);
-      StorageManager::get().add_table(table_name, std::move(table));
+      // Prepare table copy which is later used as the master to copy from.
+      _sqlite->create_table_from_tbl(table_file, table_name + _master_table_suffix);
     }
 
     opossum::Topology::use_numa_topology();
     opossum::CurrentScheduler::set(std::make_shared<opossum::NodeQueueScheduler>());
+  }
+
+ protected:
+  // Determines whether a table has been modified.
+  bool is_table_modified(const std::shared_ptr<Table> table) const {
```

Comment:
Don't have time for a full review right now, just a quick comment. Instead of crawling through the MVCC columns, which is potentially a lot of work, couldn't we just crawl the LQP for Update/Delete/Insert and invalidate the tables that those nodes reference?

```c++
visit_lqp(lqp, [&](const auto& node)) {
  if (node->type == LQPNodeType::Update) {
    StorageManager::get().remove_table(/* table that the UpdateNode references */);
  }
}
```

Reply:
Ah I knew I forgot one point you told me last week.

Will look into it!

Reply:
As you are Mr. LQP: is there no nicer way to find modifying nodes apart from a check like `type IN ('update',  'insert', ...)`?

And what would happen if we introduce UPSERTs (dunno if this is just rewritten to other operators, but let's assume it's a new operator)?

Reply:
> is there no nicer way to find modifying nodes apart from a check like type IN ('update', 'insert', ...)?

Not right now, I am afraid. The best weapon against new operators breaking existing code is an exhausting enumeration of all handled types in a switch statement. This way the compiler will error on this switch if a new type is introduced

```c++
switch (node->type) {
  case Update: case Insert: case Delete: ...

  case Every: case Other: case Type:
    // do nothing.
}

You could put the entire thing into a lqp_utils function `lqp_get_modified_tables()`